### ISSUE TITLE
Define no-extraneous-dependencies devDependencies option to files not included in the npm package

### ISSUE THREAD

**Issue by pvdlg** [import-js/eslint-plugin-import#1171]
Currently the default `devDependencies` option of [no-extraneous-dependencies](https://github.com/benmosher/eslint-plugin-import/blob/HEAD/docs/rules/no-extraneous-dependencies.md) rule is set to `false`, so any file importing a module defined as a `devDependencies` will trigger a linting error.

That's a problem for tests and build files, for example, that typically use the modules defined in `devDependencies`. So one has to specify those files like this `["error", {"devDependencies": ["build/**", "test/**"]}]`.

A possible way to determine a good default value would be to look at the files included in the npm module: files in [package.json#files](https://docs.npmjs.com/files/package.json#files) and not in `. npmignore`.

So with the following project tree:
```
project/
├── src/
│   ├── source.js
│   ├── other-source.js
├── test/
│   ├── source.test.js
│   ├── other-source.test.js
└── build.js
└── index.js
```
and `package.json#files` set to `["index.js", "src"]` we could determine that `devDependencies` option of `no-extraneous-dependencies` should be set by default to `["**/*.{js,jsx}", "index.js", "!src/**"]`.

Currently it seems the glob option doesn't support negative globs, but it can be solved easily by using [globby](https://github.com/sindresorhus/globby).

**Comment by ljharb** [comment]
The files field is dangerous and an antipattern; npmignore is preferred, so that wouldn’t be something I’d want supported here.

There’s no safe default value imo; just like we can’t assume “src” should be linted and “build” should be ignored.

**Comment by pvdlg** [comment]
> The files field is dangerous and an antipattern; npmignore is preferred, so that wouldn’t be something I’d want supported here.

Why is it dangerous? It's supported by npm, so I think the resolution what goes in the npm module should follow what npm does. It's implemented in [npm-packlist](https://github.com/npm/npm-packlist).

> There’s no safe default value imo; just like we can’t assume “src” should be linted and “build” should be ignored.

Is there any situation where you want to import dev dependencies from the code included the npm package? That would always create a runtime issue, isn't it?
And is there any situation where you don't want to prevent importing dev dependencies in the code not included in the npm package?

**Comment by ljharb** [comment]
With a blacklist ("npmignore"), if you forget to include a file, the worst damage you can cause is "an extra file is installed in the package, but everything still works". With a whitelist ("files"), the worst damage you can cause is breaking every downstream consumer. That it's a supported feature doesn't mean it's a good idea.

I'm confused; a package's dev deps simply won't be installed when it's a dependency, so npm already prevents that at runtime for us, from external code. The point of the lint rule is to ensure that you only ever import from things listed in package.json, except for the files you've explicitly indicated aren't part of the runtime behavior (like tests, etc).

**Comment by pvdlg** [comment]
> With a blacklist ("npmignore"), if you forget to include a file, the worst damage you can cause is "an extra file is installed in the package, but everything still works". With a whitelist ("files"), the worst damage you can cause is breaking every downstream consumer. That it's a supported feature doesn't mean it's a good idea.

I get the point, however a consequence of that is a lot of npm packages including a bunch of useless things at runtime such as tests files. That contribute the weight of your `node_modules`. Anyway, I'm not sure it's the place to talk about whitelist/blacklist good practice. I just think if we were to implement something it should follow what npm does as it's the de-facto standard to determine the rule related to packaging npm modules. If the consensus is that `package.json#files` is a bad practice, then npm should deprecate it and remove it in a future version.

> I'm confused; a package's dev deps simply won't be installed when it's a dependency, so npm already prevents that at runtime for us, from external code.

Yes, so if you have a file that is included in the npm package and import a dev dependency, then at runtime it will fail because the dev dependency is not included. The goal of `no-extraneous-dependencies` is to give you a linting error at dev time, rather than publishing a package that will break.

You mention earlier that there is no safe default. My point is that in this specific situation I don't see any case where the proposed default wouldn't work.
Basically it would configure the rule this way:
- Files included in npm package => Allow to import only regular dependencies (no dev dependencies)
- Files not included in npm package => Allow to import any regular or dev dependency

Is there a situation where it's not the desired behavior?

**Comment by ljharb** [comment]
The weight of node_modules is not a problem, disk space is infinite and free - bugs, however, are a problem. Certainly off topic, but I wanted to explain why relying on "files" isn't something i'd ever want in this plugin.

I agree that for the subset of packages using "files", the rule could relax the requirements in paths outside "files" - however, in the extremely common use case for "a package uses babel", `src` wouldn't be in "files" but `lib` would, and `lib` would be eslintignored anyways, and then `src` would have dev deps allowed - defeating the entire purpose of the rule.

**Comment by pvdlg** [comment]
> Certainly off topic, but I wanted to explain why relying on "files" isn't something i'd ever want in this plugin.

Yes, but maybe this rule (or another one) should also work for internals import? Like if your published code relies on a file that you forgot to include in `files` the rule would give an error.

Basically with the following:
```
project/
├── lib/
│   ├── source.js
│   ├── other-source.js
└── index.js
```
If `index.js` is:
```js
const source = require('./lib/source.js');

....
```
And you either define `"files": ["index.js"]` or `.npmignore` with `lib/**` then the rule should errors out because `index.js` which is included in the package relies on `lib/source.js` which is not.

That would be something useful and that can happen with either `files` or `.npmignore`, even the mistake is easier to make with `files`.

Another example with a the following where the files in `src` are compiled into `lib`:
```
project/
├── src/
│   ├── source.js
│   ├── other-source.js
├── lib/
│   ├── source.js
│   ├── other-source.js
└── index.js
```
And `index.js`:
If `index.js` is:
```js
const source = require('./lib/source.js');

....
```

You would add `lib` to your `.gitignore` to not push it to the repo. If you forget to create an `.npmignore` file, then `lib` will be excluded from your package and it will break at runtime.
In such case the rule would give you an error, preventing that problem to happen.

> however, in the extremely common use case for "a package uses babel", src wouldn't be in "files" but lib would, and lib would be eslintignored anyways, and then src would have dev deps allowed - defeating the entire purpose of the rule.

Yes indeed, I didn't think about that case. We would need a way to figure out that the content of `lib` (which is included in the npm package) comes from `src` and therefore `src` should be considered by the rule as if was included in the npm package (i.e. prevent dev dependencies import there).
Unfortunately I'm not aware of a good way to do that...

**Comment by ljharb** [comment]
You’re also forgetting about deep imports - using the “main” (which need not be index.js) isn’t sufficient.

The good way to do this is what we have already - each project defines precisely for itself what is ignored (eslintignore/gitignore) and what is “not runtime” (this rule’s config). The solution can’t imo be made more generic without overly complicating the config.

**Comment by jeremy-daley-kr** [comment]
Someone might've fixed negative globs since this issue. I'm on `2.26.0` and am able to do `['!*']` for turning off the `this should be a dependency instead of devDependency` thing altogether.

**Comment by ljharb** [comment]
Closing, then, unless that's not actually the case :-)
